### PR TITLE
Supermod

### PR DESCRIPTION
**Pull request by haglobah**
From the conversation today:

- [x] show full new messages count on all tabs
- [x] undo queue should have maxheight with overflow-y
- [x] reloading the page should leave you in the same collection
- [x] any indicator where in which stack I am right now
- [x] rejecting post without dm should be fine
- [x] entering the edit field should not reload the page
- [x] admins should be able to delete comments from the supermod view too

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1218572636895901) by [Unito](https://www.unito.io)

### CODE REVIEW

**Comment by haglobah** [comment]
@Discordius

**Comment by b0b3rt** [comment]
(I'm not sure what's happening here but I don't think this PR should be changing 60 files; I would be surprised if it needed more than 20 and that's including 5+ generated file diffs)

**Comment by b0b3rt** [comment]
Also I'm not sure what "admins should be able to delete comments from the supermod view too" means; we don't delete anything from supermod (or remove anything from it in a way that seems analogous to deletion)

**Comment by haglobah** [comment]
The 60 files are something like:

- 6 small features touching ~2-10 files each
- a bit over 20 files that are fixing the bash invocations in the scripts to be distro-agnostic
- 2 nix files (which I can a bit unhappily, but easily keep out)
- 1 md file, which is an oversight of mine

**Comment by haglobah** [comment]
re: Should be able to delete comments: Prior to this PR, the supermod per-user view of a comment doesn't give you the option to delete that same comment (you need to go to the comment in the context of the post to have that shown)

**Comment by b0b3rt** [comment]
Brief notes:
- need to have a much higher bar for LLM-generated tests (almost none of these things were bugs, and the one thing that was a bug doesn't seem that likely to regress with future changes; basically assume no unless you've looked at each individual test and explicitly thought about whether it's the kind of test that would ever actually fail in practice)
- prefer no random new modules that export functions/etc that are only imported in a single place unless there's some very clear conceptual boundary being established (i.e. a separate file for a 3rd-party API client is fine even if there's only one consumer so far)
- I don't think a `skipRejectionPM` field is the idiomatic way to handle that kind of thing in this codebase.  Probably should make a new custom resolver for rejecting posts which accepts an optional flag, rather than introducing an   (Separately, what's the UI for that like?)
- whatever is going on with `userGqlQueryHandlers` does not look remotely kosher

I haven't examined the rest of the code in the PR very closely but I am worried that the process that generated the `userGqlQueryHandlers` change also generated the rest of the code

File: packages/lesswrong/server/repos/UsersRepo.ts [hunk 1]
```diff
@@ -51,6 +55,31 @@ class UsersRepo extends AbstractRepo<"Users"> {
     super(Users);
   }
 
+  async getNewUsersByOldestUnreviewedContent(selector: MongoSelector<DbUser>, limit: number, offset: number): Promise<DbUser[]> {
+    const usersQuery = new SelectQuery(this.getCollection().getTable(), selector).compile();
+    const postsQuery = new SelectQuery(Posts.getTable(), unreviewedUserPostSelector).compile(usersQuery.args.length);
+    const commentsQuery = new SelectQuery(Comments.getTable(), unreviewedUserCommentSelector).compile(usersQuery.args.length + postsQuery.args.length);
```
Comment (b0b3rt) [inline packages/lesswrong/server/repos/UsersRepo.ts:66]: you do not see my 😱 face from here

Comment (b0b3rt) [reply packages/lesswrong/server/repos/UsersRepo.ts:66]: (also, this will not work very correctly.  there are many users with unreviewed content who have been removed from queue.)

Comment (haglobah) [reply packages/lesswrong/server/repos/UsersRepo.ts:66]: > you do not see my 😱 face from here

Please explain :) (I now very little about gql, but it doesn't look totally crazy?)

File: packages/lesswrong/server/resolvers/moderationUserQueueCounts.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import { UsersViews } from '@/lib/collections/users/views';
+import { getUserReviewGroup } from '@/lib/collections/users/reviewGroupResolvers';
+import { viewTermsToQuery } from '@/lib/utils/viewUtils';
+import { userIsAdminOrMod } from '@/lib/vulcan-users/permissions';
+
+export async function getModerationUserQueueCounts(context: ResolverContext): Promise<Record<ReviewGroup, number>> {
```
Comment (b0b3rt) [inline packages/lesswrong/server/resolvers/moderationUserQueueCounts.ts]: I don't think I love this implementation, it feels very "shortest path from the previous implementation with no regard for what a good first-principles design would look like" to me. (LLMs!)

Maybe our logic for determining which review group a user should be in is complicated enough that there's not much we can do about it... man, I knew the offboarding tab was a mistake.

File: packages/lesswrong/components/sunshineDashboard/supermod/ModerationContentList.tsx [hunk 3]
```diff
@@ -28,6 +34,30 @@ const styles = defineStyles('ModerationContentList', (theme: ThemeType) => ({
...
+    padding: '2px 6px',
+    borderRadius: 4,
+    backgroundColor: theme.palette.greyAlpha(0.1),
+    '&[data-queue="newContent"]': {
+      backgroundColor: theme.palette.panelBackground.sunshineNewPosts,
+    },
+    '&[data-queue="maybeSpam"], &[data-queue="offboard"]': {
+      backgroundColor: theme.palette.panelBackground.sunshineWarningHighlight,
+    },
+    '&[data-queue="highContext"], &[data-queue="automod"], &[data-queue="snoozeExpired"]': {
```
Comment (b0b3rt) [inline packages/lesswrong/components/sunshineDashboard/supermod/ModerationContentList.tsx]: Data attributes are the idiomatic way to apply styles to individual spans.  Maybe I'm missing something; is there any reason these can't just be individual style classes?

File: packages/lesswrong/components/sunshineDashboard/supermod/ModerationInbox.tsx [hunk 4]
```diff
@@ -35,6 +37,15 @@ import ModerationUndoHistory from './ModerationUndoHistory';
 // depends on whether the opened user is already in the users list.)
 const ModerationInboxDataQuery = gql(`
   query ModerationInboxDataQuery($userSelector: UserSelector, $postSelector: PostSelector, $classifiedPostSelector: PostSelector, $userLimit: Int, $postLimit: Int, $curationLimit: Int) {
+    moderationUserQueueCounts {
```
Comment (b0b3rt) [inline packages/lesswrong/components/sunshineDashboard/supermod/ModerationInbox.tsx:40]: If we can live with just having a total count of all users in the queue, which can be shown on the `All` tab (and not having un-truncated counts for each individual group) we can just return `totalCount` from the `users` query below.  (That might be expensive to load, though; not sure.  But if it is, then I'm also worried about the performance implications of this new resolver.)

File: packages/lesswrong/components/sunshineDashboard/supermod/ModerationInbox.tsx [hunk 5]
```diff
@@ -560,8 +573,11 @@ const ModerationInbox = () => {
     );
   }
 
+  if (!data?.moderationUserQueueCounts) throw new Error('Unable to load moderation queue counts');
```
Comment (b0b3rt) [inline packages/lesswrong/components/sunshineDashboard/supermod/ModerationInbox.tsx]: We generally don't throw from components

File: packages/lesswrong/components/sunshineDashboard/supermod/ModerationInbox.tsx (hunk 4 above)
Comment (haglobah) [reply packages/lesswrong/components/sunshineDashboard/supermod/ModerationInbox.tsx:40]: Hmm, the idea was that one can see for every tab how many are currently loaded and still not loaded.
(I didn't have a noticeable perf change, so I guess that might be fine?
case in point would be that the numbers are actually quite small, like currently on local I had ~600 users in 'all', and that is I already higher as it should get?)

File: packages/lesswrong/server/resolvers/moderationUserQueueCounts.ts (hunk 2 above)
Comment (haglobah) [reply packages/lesswrong/server/resolvers/moderationUserQueueCounts.ts]: Yeah, from the little context I have I'm on the "this is even more complicated otherwise" side I think

File: packages/lesswrong/server/collections/posts/mutations.ts [hunk 6]
```diff
@@ -199,7 +199,12 @@ export async function createPost({ data }: { data: CreatePostDataInput & { _id?:
   return documentWithId;
 }
 
-export async function updatePost({ selector, data }: { data: UpdatePostDataInput | Partial<DbPost>; selector: SelectorInput }, context: ResolverContext) {
+export interface UpdatePostOptions {
+  // Reject without messaging the author; only the rejectPost mutation sets this.
+  skipRejectionPM?: boolean;
```
Comment (b0b3rt) [inline packages/lesswrong/server/collections/posts/mutations.ts:204]: Sorry to keep nitpicking, but: there is a pretty strong convention about how "default" resource mutation functions are defined, and if it turns out that we have behavior which justifies writing a custom resolver for rather than using the default mutation function directly, that behavior should be executed in the custom resolver, not made the responsibility of the default mutation function (which should still be used to handle actually updated the relevant fields).  Previously, we were always doing it by default when posts were rejected, so it was appropriate for it to live in updatePost.

Can discuss more in person; I'm not totally happy with my reasoning here
